projects/erigon_riscv_proof_sourcing.md

```diff
@@ -0,0 +1,148 @@
+
+### First iteration of the transpilation module (Weeks 5-10)
+The first few weeks will involve us writing the first iteration of the transpilation module. This involves handling the instructions mapping and handling the first version of ELF binary generation.
+
+**Delivery:** We should after this be able to prove simple contracts that we have written ourself, we will also still be in control of the environment. There might be no support for 256bit arithmetic operations, some limits to control flow needs or similar simplifications.
+
+### Stabilizing the transpilation module (Weeks 11-17)
+For the next milestone we want to stabilize and solve some of the [possible challenges](#possible-challenges) to improve our support of real world contracts.
+
+**Delivery:** We should be able to transpile Uniswap V2 and other [real world applications](https://etherscan.io/topstat#HotContracts) and operate in close to a real world environment.
+
+### Optimize, benchmarking and evaluation (Weeks 18-21)
+At this point we hope to have an implementation that we can run experiments on to get a sense of the feasibility of this approach. We want to benchmark this against other existing approaches. The result of this should be published as part of the conclusion of this project.
+
+**Delivery**: We should provide a quantitative performance comparison between our transpilation approach and existing alternatives including metrics such as transpilation time, binary size, and execution speed.
+
+## Possible challenges
+This is a big project and there will obviously be some challenges along the way, we list some of them with possible mitigations that we will try to implement.
+
+### ISA edge case mapping
+- **Challenge**: Different ISAs handle edge cases differently (numeric overflows, memory modeling, instruction errors)
+- **Impact**: Critical - security issues from incorrect transpilation
+- **Mitigation**: Test suite comparing outputs between environments
+- **Timeline**: Ongoing throughout the project
+
+### Transpilation correctness
+- **Challenge**: Bugs in transpilation module pose serious security risks
+- **Impact**: Critical - incorrect translation could compromise system security
+- **Mitigation**: Testing, consider [formal verification methods](https://verified-zkevm.org/)
+- **Timeline**: Continuous verification as opcodes are implemented
+
+### Word size
+- **Challenge**: RISC-V uses 32/64-bit words, EVM uses 256-bit
+- **Impact**: High - affects all arithmetic operations
+- **Mitigation**: There are existing algorithms for supporting arbitrary-precision arithmetic which we can implement in RISC-V.
+- **Timeline**: Weeks 11-17
+
+### Special opcodes and Precompiles
+- **Challenge**: Some opcodes will be harder to implement directly. For instance `keccak256` and some of the state introspection opcodes like `EXTCODESIZE` / `SELFBALANCE` / `CALLDATALOAD`. Precompiles also don't have a direct transpilation mapping as they are syscall- like.
```

Comment:
One related thing I am curious about whether dynamic jumps (JUMP, JUMPI) also pose a challenge?

Reply:
If you are only transpiling the executed path at runtime, I imagine it's only prove of a single execution trace, not the whole contract

Reply:
Yeah, that is correct, it's not really an issue since we only transpile the executed path. 

I still think supporting JUMP(I) would be interesting as it could maybe allow for optimizations as we don't need to compile for each transaction (although the output binary will be larger and more complex so might be some threshold). 

Added a note on it in 336d5494e885665472b07516d680d524586317b4